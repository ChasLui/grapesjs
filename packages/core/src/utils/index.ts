import Dragger from './Dragger';
import Sorter from './Sorter';
import Resizer from './Resizer';
import * as mixins from './mixins';
import { Module } from '../abstract';
import EditorModel from '../editor/model/Editor';
import ComponentSorter from './ComponentSorter';
import StyleManagerSorter from './StyleManagerSorter';

export default class UtilsModule extends Module {
  Sorter = Sorter;
  Resizer = Resizer;
  Dragger = Dragger;
  ComponentSorter = ComponentSorter;
  StyleManagerSorter = StyleManagerSorter;
  helpers = { ...mixins };

  constructor(em: EditorModel) {
    super(em, 'Utils');
  }

  destroy() {}
}
