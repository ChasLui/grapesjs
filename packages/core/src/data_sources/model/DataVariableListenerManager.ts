import { DataSourcesEvents, DataVariableListener } from '../types';
import { stringToPath } from '../../utils/mixins';
import { Model } from '../../common';
import EditorModel from '../../editor/model/Editor';
import DataVariable from './DataVariable';
import ComponentView from '../../dom_components/view/ComponentView';
import ComponentDataVariable from './ComponentDataVariable';

export interface DataVariableListenerManagerOptions {
  model: Model | ComponentView;
  em: EditorModel;
  dataVariable: DataVariable | ComponentDataVariable;
  updateValueFromDataVariable: (value: any) => void;
}

export default class DataVariableListenerManager {
  private dataListeners: DataVariableListener[] = [];
  private em: EditorModel;
  private model: Model | ComponentView;
  private dataVariable: DataVariable | ComponentDataVariable;
  private updateValueFromDataVariable: (value: any) => void;

  constructor(options: DataVariableListenerManagerOptions) {
    this.em = options.em;
    this.model = options.model;
    this.dataVariable = options.dataVariable;
    this.updateValueFromDataVariable = options.updateValueFromDataVariable;

    this.listenToDataVariable();
  }

  private onChange = () => {
    const value = this.dataVariable.getDataValue();
    this.updateValueFromDataVariable(value);
  };

  listenToDataVariable() {
    const { em, dataVariable, model } = this;
    const { path } = dataVariable.attributes;
    const normPath = stringToPath(path || '').join('.');
    const [ds, dr] = this.em.DataSources.fromPath(path);

    this.removeListeners();

    const dataListeners: DataVariableListener[] = [];
    ds && dataListeners.push({ obj: ds.records, event: 'add remove reset' });
    dr && dataListeners.push({ obj: dr, event: 'change' });
    dataListeners.push(
      { obj: dataVariable, event: 'change:path change:defaultValue' },
      { obj: em.DataSources.all, event: 'add remove reset' },
      { obj: em, event: `${DataSourcesEvents.path}:${normPath}` },
    );

    dataListeners.forEach((ls) => model.listenTo(ls.obj, ls.event, this.onChange));

    this.dataListeners = dataListeners;
  }

  private removeListeners() {
    const { model } = this;
    this.dataListeners.forEach((ls) => model.stopListening(ls.obj, ls.event, this.onChange));
    this.dataListeners = [];
  }
}
