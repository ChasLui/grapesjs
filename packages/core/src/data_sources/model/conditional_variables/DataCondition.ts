import { NumberOperation } from './operators/NumberOperator';
import { StringOperation } from './operators/StringOperations';
import { GenericOperation } from './operators/GenericOperator';
import { Model } from '../../../common';
import { LogicalOperation } from './operators/LogicalOperator';
import DynamicVariableListenerManager from '../DataVariableListenerManager';
import EditorModel from '../../../editor/model/Editor';
import { Condition } from './Condition';
import DataVariable, { DataVariableDefinition } from '../DataVariable';
import { evaluateVariable, isDataVariable } from '../utils';

export const ConditionalVariableType = 'conditional-variable';
export type Expression = {
  left: any;
  operator: GenericOperation | StringOperation | NumberOperation;
  right: any;
};

export type LogicGroup = {
  logicalOperator: LogicalOperation;
  statements: (Expression | LogicGroup | boolean)[];
};

export type ConditionalVariableDefinition = {
  type: typeof ConditionalVariableType;
  condition: Expression | LogicGroup | boolean;
  ifTrue: any;
  ifFalse: any;
};

export class DataCondition extends Model {
  lastEvaluationResult: boolean;
  private condition: Condition;
  private em: EditorModel;
  private variableListeners: DynamicVariableListenerManager[] = [];
  private _onValueChange?: () => void;

  defaults() {
    return {
      type: ConditionalVariableType,
      condition: false,
    };
  }

  constructor(
    condition: Expression | LogicGroup | boolean,
    public ifTrue: any,
    public ifFalse: any,
    opts: { em: EditorModel; onValueChange?: () => void },
  ) {
    if (typeof condition === 'undefined') {
      throw new MissingConditionError();
    }

    super();
    this.condition = new Condition(condition, { em: opts.em });
    this.em = opts.em;
    this.lastEvaluationResult = this.evaluate();
    this.listenToDataVariables();
    this._onValueChange = opts.onValueChange;
  }

  evaluate() {
    return this.condition.evaluate();
  }

  getDataValue(): any {
    return this.lastEvaluationResult ? evaluateVariable(this.ifTrue, this.em) : evaluateVariable(this.ifFalse, this.em);
  }

  reevaluate(): void {
    this.lastEvaluationResult = this.evaluate();
  }

  set onValueChange(newFunction: () => void) {
    this._onValueChange = newFunction;
    this.listenToDataVariables();
  }

  private listenToDataVariables() {
    if (!this.em) return;

    // Clear previous listeners to avoid memory leaks
    this.cleanupListeners();

    const dataVariables = this.getDependentDataVariables();

    dataVariables.forEach((variable) => {
      const variableInstance = new DataVariable(variable, { em: this.em });
      const listener = new DynamicVariableListenerManager({
        model: this as any,
        em: this.em!,
        dataVariable: variableInstance,
        updateValueFromDataVariable: (() => {
          this.reevaluate();
          this._onValueChange?.();
        }).bind(this),
      });

      this.variableListeners.push(listener);
    });
  }

  getDependentDataVariables() {
    const dataVariables: DataVariableDefinition[] = this.condition.getDataVariables();
    if (isDataVariable(this.ifTrue)) dataVariables.push(this.ifTrue);
    if (isDataVariable(this.ifFalse)) dataVariables.push(this.ifFalse);

    return dataVariables;
  }

  private cleanupListeners() {
    this.variableListeners.forEach((listener) => listener.destroy());
    this.variableListeners = [];
  }

  toJSON() {
    return {
      type: ConditionalVariableType,
      condition: this.condition,
      ifTrue: this.ifTrue,
      ifFalse: this.ifFalse,
    };
  }
}
export class MissingConditionError extends Error {
  constructor() {
    super('No condition was provided to a conditional component.');
  }
}
