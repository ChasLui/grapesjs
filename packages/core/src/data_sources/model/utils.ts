import EditorModel from '../../editor/model/Editor';
import { DynamicValue, DynamicValueDefinition } from '../types';
import { ConditionalVariableType, DataCondition } from './conditional_variables/DataCondition';
import DataVariable, { DataVariableType } from './DataVariable';

export function isDynamicValueDefinition(value: any): value is DynamicValueDefinition {
  return typeof value === 'object' && [DataVariableType, ConditionalVariableType].includes(value.type);
}

export function isDynamicValue(value: any): value is DynamicValue {
  return value instanceof DataVariable || value instanceof DataCondition;
}

export function isDataVariable(variable: any) {
  return variable?.type === DataVariableType;
}

export function isDataCondition(variable: any) {
  return variable?.type === ConditionalVariableType;
}

export function evaluateVariable(variable: any, em: EditorModel) {
  return isDataVariable(variable) ? new DataVariable(variable, { em }).getDataValue() : variable;
}
