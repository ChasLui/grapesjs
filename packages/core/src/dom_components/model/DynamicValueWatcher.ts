import DynamicVariableListenerManager from '../../data_sources/model/DataVariableListenerManager';
import { evaluateDynamicValueDefinition, isDynamicValueDefinition } from '../../data_sources/model/utils';
import EditorModel from '../../editor/model/Editor';

export class DynamicValueWatcher {
  dynamicVariableListeners: { [key: string]: DynamicVariableListenerManager } = {};
  constructor(
    private updateFn: (key: string, value: any) => void,
    private em: EditorModel,
  ) {}

  static getStaticValues(
    values: {
      [key: string]: any;
    },
    em: EditorModel,
  ) {
    const evaluatedValues = { ...values };
    const propsKeys = Object.keys(evaluatedValues);
    for (let index = 0; index < propsKeys.length; index++) {
      const key = propsKeys[index];
      if (!isDynamicValueDefinition(evaluatedValues[key])) continue;
      const { value } = evaluateDynamicValueDefinition(evaluatedValues[key], em);
      evaluatedValues[key] = value;
    }

    return evaluatedValues;
  }

  static areStaticValues(values: { [key: string]: any }) {
    return Object.keys(values).every((key) => {
      return !isDynamicValueDefinition(values[key]);
    });
  }

  watchDynamicValue(values: { [key: string]: any }) {
    const dynamicProps = this.getDynamicValues(values);
    const propsKeys = Object.keys(dynamicProps);
    for (let index = 0; index < propsKeys.length; index++) {
      const key = propsKeys[index];

      this.dynamicVariableListeners[key] = new DynamicVariableListenerManager({
        em: this.em,
        dataVariable: dynamicProps[key],
        updateValueFromDataVariable: (value: any) => {
          this.updateFn.bind(this)(key, value);
        },
      });
    }

    return dynamicProps;
  }

  private getDynamicValues(values: { [key: string]: any }) {
    const dynamicValues = { ...values };
    const propsKeys = Object.keys(dynamicValues);
    for (let index = 0; index < propsKeys.length; index++) {
      const key = propsKeys[index];
      if (!isDynamicValueDefinition(dynamicValues[key])) {
        delete dynamicValues[key];
        continue;
      }
      const { variable } = evaluateDynamicValueDefinition(dynamicValues[key], this.em);
      dynamicValues[key] = variable;
    }

    return dynamicValues;
  }

  /**
   * removes listeners to stop watching for changes,
   * if keys argument is omitted, remove all listeners
   * @argument keys
   */
  removeListeners(keys?: string[]) {
    const propsKeys = keys ? keys : Object.keys(this.dynamicVariableListeners);
    propsKeys.forEach((key) => this.dynamicVariableListeners[key].destroy());
  }
}
