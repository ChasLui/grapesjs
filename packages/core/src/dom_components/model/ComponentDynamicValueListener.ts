import { ObjectAny } from '../../common';
import EditorModel from '../../editor/model/Editor';
import Component from './Component';
import { DynamicValueWatcher } from './DynamicValueWatcher';

export class ComponentDynamicValueListener {
  propertyWatchClass: DynamicValueWatcher;
  attributeWatchClass: DynamicValueWatcher;

  constructor(
    private component: Component,
    em: EditorModel,
  ) {
    this.propertyWatchClass = new DynamicValueWatcher((key: string, value: any) => {
      this.component.set(key, value);
    }, em);

    this.attributeWatchClass = new DynamicValueWatcher((key: string, value: any) => {
      this.component.setAttributes({ [key]: value });
    }, em);
  }

  static evaluateComponentDef(values: ObjectAny, em: EditorModel) {
    const props = DynamicValueWatcher.getStaticValues(values, em);
    if (values.attributes) {
      props.attributes = DynamicValueWatcher.getStaticValues(values.attributes, em);
    }

    return props;
  }

  watchComponentDef(values: ObjectAny) {
    this.watchProps(values);
    this.watchAttributes(values.attributes);
  }

  watchProps(props: ObjectAny) {
    this.propertyWatchClass.watchDynamicValue(props);
  }

  getDynamicPropsDefs() {
    return this.propertyWatchClass.getAllSerializableValues();
  }

  setAttributes(attributes: ObjectAny) {
    this.attributeWatchClass.removeListeners();
    this.attributeWatchClass.watchDynamicValue(attributes);
  }

  watchAttributes(attributes: ObjectAny) {
    this.attributeWatchClass.watchDynamicValue(attributes);
  }

  removeAttributes(attributes: string[]) {
    this.attributeWatchClass.removeListeners(attributes);
  }

  getAttributesDefsOrValues(attributes: ObjectAny) {
    return this.attributeWatchClass.getSerializableValues(attributes);
  }
}
