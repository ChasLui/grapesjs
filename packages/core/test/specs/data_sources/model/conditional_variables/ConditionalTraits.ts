import { DataSourceManager, Editor } from '../../../../../src';
import { DataVariableType } from '../../../../../src/data_sources/model/DataVariable';
import { MissingConditionError } from '../../../../../src/data_sources/model/conditional_variables/DataCondition';
import { ConditionalVariableType } from '../../../../../src/data_sources/model/conditional_variables/DataCondition';
import { GenericOperation } from '../../../../../src/data_sources/model/conditional_variables/operators/GenericOperator';
import { NumberOperation } from '../../../../../src/data_sources/model/conditional_variables/operators/NumberOperator';
import Component from '../../../../../src/dom_components/model/Component';
import ComponentWrapper from '../../../../../src/dom_components/model/ComponentWrapper';
import EditorModel from '../../../../../src/editor/model/Editor';
import { filterObjectForSnapshot, setupTestEditor } from '../../../../common';

describe('TraitConditionalVariable', () => {
  let editor: Editor;
  let em: EditorModel;
  let dsm: DataSourceManager;
  let cmpRoot: ComponentWrapper;

  beforeEach(() => {
    ({ editor, em, dsm, cmpRoot } = setupTestEditor());
  });

  afterEach(() => {
    em.destroy();
  });

  it('should add a trait with a condition evaluating to a string', () => {
    const component = cmpRoot.append({
      tagName: 'h1',
      type: 'text',
      traits: [
        {
          type: 'text',
          name: 'title',
          value: {
            type: ConditionalVariableType,
            condition: {
              left: 0,
              operator: NumberOperation.greaterThan,
              right: -1,
            },
            ifTrue: 'Some title',
          },
        },
      ],
    })[0];

    testComponentAttr(component, 'title', 'Some title');
  });

  it('should add a trait with a data-source condition', () => {
    const dataSource = {
      id: 'ds1',
      records: [{ id: 'left_id', left: 'Name1' }],
    };
    dsm.add(dataSource);

    const component = cmpRoot.append({
      tagName: 'h1',
      type: 'text',
      traits: [
        {
          type: 'text',
          name: 'title',
          value: {
            type: ConditionalVariableType,
            condition: {
              left: {
                type: DataVariableType,
                path: 'ds1.left_id.left',
              },
              operator: GenericOperation.equals,
              right: 'Name1',
            },
            ifTrue: 'Valid name',
            ifFalse: 'Invalid name',
          },
        },
      ],
    })[0];

    testComponentAttr(component, 'title', 'Valid name');
  });

  it('should change trait value with changing data-source value', () => {
    const dataSource = {
      id: 'ds1',
      records: [{ id: 'left_id', left: 'Name1' }],
    };
    dsm.add(dataSource);

    const component = cmpRoot.append({
      tagName: 'h1',
      type: 'text',
      traits: [
        {
          type: 'text',
          name: 'title',
          value: {
            type: ConditionalVariableType,
            condition: {
              left: {
                type: DataVariableType,
                path: 'ds1.left_id.left',
              },
              operator: GenericOperation.equals,
              right: 'Name1',
            },
            ifTrue: 'Correct name',
            ifFalse: 'Incorrect name',
          },
        },
      ],
    })[0];

    testComponentAttr(component, 'title', 'Correct name');
    dsm.get('ds1').getRecord('left_id')?.set('left', 'Different name');
    testComponentAttr(component, 'title', 'Incorrect name');
  });

  it('should throw an error if no condition is passed in trait', () => {
    expect(() => {
      cmpRoot.append({
        tagName: 'h1',
        type: 'text',
        traits: [
          {
            type: 'text',
            name: 'invalidTrait',
            value: {
              type: ConditionalVariableType,
            },
          },
        ],
      });
    }).toThrow(MissingConditionError);
  });

  // TODO: Should we keep or delete saving triats with dynamic values?
  it.skip('should store traits with conditional values correctly', () => {
    const conditionalTrait = {
      type: ConditionalVariableType,
      condition: {
        left: 0,
        operator: NumberOperation.greaterThan,
        right: -1,
      },
      ifTrue: 'Positive',
    };
    cmpRoot.append({
      tagName: 'h1',
      type: 'text',
      traits: [
        {
          type: 'text',
          name: 'dynamicTrait',
          value: conditionalTrait,
        },
      ],
    })[0];

    const projectData = editor.getProjectData();
    const snapshot = filterObjectForSnapshot(projectData);
    expect(snapshot).toMatchSnapshot(``);
    const page = projectData.pages[0];
    const frame = page.frames[0];
    const storedComponent = frame.component.components[0];

    // expect(storedComponent[dynamicAttrKey]).toEqual({
    //   dynamicTrait: conditionalTrait,
    // });
  });

  // TODO: Should we keep or delete saving triats with dynamic values?
  it.skip('should load traits with conditional values correctly', () => {
    const projectData = {
      pages: [
        {
          frames: [
            {
              component: {
                components: [
                  {
                    attributes: {
                      dynamicTrait: 'Default',
                    },
                    // [dynamicAttrKey]: {
                    //   dynamicTrait: {
                    //     condition: {
                    //       left: 0,
                    //       operator: '>',
                    //       right: -1,
                    //     },
                    //     ifTrue: 'Positive',
                    //     type: 'conditional-variable',
                    //   },
                    // },
                    type: 'text',
                  },
                ],
                type: 'wrapper',
              },
            },
          ],
          type: 'main',
        },
      ],
    };

    editor.loadProjectData(projectData);
    const components = editor.getComponents();
    const component = components.models[0];
    expect(component.getAttributes()).toEqual({ dynamicTrait: 'Positive' });
  });

  it('should be property on the component with `changeProp:true`', () => {
    const dataSource = {
      id: 'ds1',
      records: [{ id: 'left_id', left: 'Name1' }],
    };
    dsm.add(dataSource);

    const component = cmpRoot.append({
      tagName: 'h1',
      type: 'text',
      traits: [
        {
          type: 'text',
          name: 'title',
          changeProp: true,
          value: {
            type: ConditionalVariableType,
            condition: {
              left: {
                type: DataVariableType,
                path: 'ds1.left_id.left',
              },
              operator: GenericOperation.equals,
              right: 'Name1',
            },
            ifTrue: 'Correct name',
            ifFalse: 'Incorrect name',
          },
        },
      ],
    })[0];

    // TODO: make dynamic values not to change the attributes if `changeProp:true`
    // expect(component.getView()?.el.getAttribute('title')).toBeNull();
    expect(component.get('title')).toBe('Correct name');

    dsm.get('ds1').getRecord('left_id')?.set('left', 'Different name');
    // expect(component.getView()?.el.getAttribute('title')).toBeNull();
    expect(component.get('title')).toBe('Incorrect name');
  });

  it('should handle objects as traits (other than dynamic values)', () => {
    const traitValue = {
      type: 'UNKNOWN_TYPE',
      condition: "This's not a condition",
      value: 'random value',
    };

    const component = cmpRoot.append({
      tagName: 'h1',
      type: 'text',
      traits: [
        {
          type: 'text',
          name: 'title',
          value: traitValue,
        },
      ],
    })[0];
    expect(component.getTrait('title').get('value')).toEqual(traitValue);
    expect(component.getAttributes().title).toEqual(traitValue);
  });
});

function testComponentAttr(component: Component, trait: string, value: string) {
  expect(component).toBeDefined();
  expect(component.getTrait(trait).get('value')).toBe(value);
  expect(component.getAttributes()[trait]).toBe(value);
  expect(component.getView()?.el.getAttribute(trait)).toBe(value);
}
