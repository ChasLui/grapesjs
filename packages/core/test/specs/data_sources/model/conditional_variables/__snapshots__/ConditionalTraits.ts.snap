// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TraitConditionalVariable should store traits with conditional values correctly 1`] = `
{
  "assets": [],
  "dataSources": [],
  "pages": [
    {
      "frames": [
        {
          "component": {
            "components": [
              {
                "attributes": {
                  "dynamicTrait": "Positive",
                },
                "attributes-dynamic-value": {
                  "dynamicTrait": {
                    "condition": {
                      "left": 0,
                      "operator": ">",
                      "right": -1,
                    },
                    "ifTrue": "Positive",
                    "type": "conditional-variable",
                  },
                },
                "tagName": "h1",
                "type": "text",
              },
            ],
            "docEl": {
              "tagName": "html",
            },
            "head": {
              "type": "head",
            },
            "stylable": [
              "background",
              "background-color",
              "background-image",
              "background-repeat",
              "background-attachment",
              "background-position",
              "background-size",
            ],
            "type": "wrapper",
          },
          "id": "data-variable-id",
        },
      ],
      "id": "data-variable-id",
      "type": "main",
    },
  ],
  "styles": [],
  "symbols": [],
}
`;
